src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs

```diff
@@ -73,13 +73,13 @@ private void EnsureCache()
                 foreach (var viewDescriptor in GetViewDescriptors(_applicationManager))
                 {
                     PageRouteModel model = null;
-                    if (viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
+                    if (_pagesOptions.AllowAreas && viewDescriptor.RelativePath.StartsWith(areaRootDirectory, StringComparison.OrdinalIgnoreCase))
```

Comment:
The fact that we're relying on the relationship between arearoot and root here isn't obvious. This deserves a comment that the ordering is significant.